[vulkan] Add hardswish and hardsigmoid activations to Vulkan

SS-JIA opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53362 [vulkan] Add hardswish and hardsigmoid activations to Vulkan**

Differential Revision: [D27098430](https://our.internmc.facebook.com/intern/diff/D27098430)

Commit:
[vulkan] Add hardswish and hardsigmoid activations to Vulkan

[ghstack-poisoned]

SS-JIA commented on aten/src/ATen/native/vulkan/glsl/hardsigmoid_.glsl:
```diff
@@ -0,0 +1,25 @@
+  const ivec3 pos = ivec3(gl_GlobalInvocationID);
+
+  if (all(lessThan(pos, uBlock.size.xyz))) {
+    vec4 outval = imageLoad(uOutput, pos)/6.0 + 0.5;
```
Not sure if something like `*(1.0/6.0)` would work better?

Commit:
Update on "[vulkan] Add hardswish and hardsigmoid activations to Vulkan"




[ghstack-poisoned]

Commit:
Update on "[vulkan] Add hardswish and hardsigmoid activations to Vulkan"




[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/glsl/hardsigmoid.glsl:
```diff
@@ -0,0 +1,23 @@
+  const ivec3 pos = ivec3(gl_GlobalInvocationID);
+
+  if (all(lessThan(pos, uBlock.size.xyz))) {
+    const vec4 outval = texelFetch(uInput, pos, 0)/6.0 + 0.5;
```
It is useful to form the habit to add the f suffix to numeric literals to indicate a single precision floating point designation (6.0f vs 6.0) as opposed to double precision since floating point literals are double precision by default.  It is not going to make a difference here but sometimes that is the difference between calling a function that has both the single and double precision overloads (say, if a div function had both float and double overloads), or evaluating a statement at the higher double precision when single was enough.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/glsl/hardswish.glsl:
```diff
@@ -0,0 +1,26 @@
+    const vec4 inval = texelFetch(uInput, pos, 0);
+    const vec4 mask1 = vec4(lessThan(inval, vec4(3.0)));
+    const vec4 mask2 = vec4(greaterThan(inval, vec4(-3.0)));
+    const vec4 outval = mask2*(mask1*(inval*(inval + 3.0)/6.0) + (1 - mask1)*inval);
```
Is it beneficial to factor out inval out of [statement a] + [statement b]?  Will that save a multiplication?

SS-JIA replied:
ah, good point

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/Clamp.cpp:
```diff
@@ -162,11 +273,31 @@ Tensor& relu_(Tensor& self) {
+  m.impl("hardswish_", hardswish_);
   m.impl("hardtanh", hardtanh);
   m.impl("hardtanh_", hardtanh_);
   m.impl("relu", relu);
```
Nice, the op list is beefing up. :)

AshkanAliabadi reviewed (approved):
Excellent, thanks!

Commit:
Update on "[vulkan] Add hardswish and hardsigmoid activations to Vulkan"


Differential Revision: [D27098430](https://our.internmc.facebook.com/intern/diff/D27098430)

[ghstack-poisoned]
